Initial work (#1)
vertx-howtos/protobuf-eventbus-howto: README.adoc

= Exchanging generated Protobuf classes on the Event Bus
:page-permalink: /
:page-github: vertx-howtos/protobuf-eventbus-howto

> This document will show you how to exchange generated Protobuf classes on the Event Bus.

Reviewer: I'd add a hyperlink to protocol buffers.

== What you will build

You will build an application that periodically generates a greeting message.
The application consists in:

* a sender verticle which sends a `GreetingRequest`
* a receiver verticle which replies to requests with a `GreetingResponse`

> == What you need

Reviewer: The Maven plugin is battery included, right?

Author: What do you mean?

Reviewer: Just making sure you don't have to install anything else to compile protobuf, etc.

Author: Then the answer is yes, batteries included.

* A text editor or an IDE,
* Java 11 or higher

== Create a project

Browse to https://start.vertx.io.
Click on advanced options to expand the hidden panel and then change the value of the following fields:

* _Group Id_: set to `io.vertx.howtos`
* _Artifact Id_: set to `protobuf-eventbus-howto`
* _Dependencies_: add `Hazelcast Cluster Manager`
* _Package_: set to `io.vertx.howtos.protobuf.eventbus`

> When you're done, click on _Generate Project_ and extract the generated archive content somewhere on your disk.

Reviewer: ```suggestion
When you're done, click on _Generate Project_ and extract the generated archive content somewhere on your filesystem.
```

[TIP]
====

> Alternatively, you may do this from the command-line:

Reviewer: ```suggestion
You may alternatively do this from the command line:
```

[source,shell]
----
curl -G https://start.vertx.io/starter.zip -d "groupId=io.vertx.howtos" -d "artifactId=protobuf-eventbus-howto" -d "packageName=io.vertx.howtos.protobuf.eventbus" -d "vertxDependencies=vertx-hazelcast" -d "jdkVersion=11" -d "buildTool=maven" --output protobuf-eventbus-howto.zip
unzip -d protobuf-eventbus-howto protobuf-eventbus-howto.zip
----
====

Before coding, we need to make some adjustments to the build file:

> * configuring a custom Vert.x `Launcher` class (that will be used as entry point when running the executable JAR)

Reviewer: ```suggestion
* configure a custom Vert.x `Launcher` class (that will be used as entry point when running the executable JAR)
```

> * adding a dependency to Protocol Buffers

Reviewer: ```suggestion
* add a dependency to _Protocol Buffers_
```

> * configuring Maven plugins to generate message classes from a `.proto` file.

Reviewer: ```suggestion
* configure Maven plugins to generate message classes from a `.proto` file.
```

Here is the content of the `pom.xml` file you should be using:

[source,xml,role="collapsed"]
.Maven `pom.xml`
----
include::pom.xml[]
----

== Implementation of the application

=== Definition of the messages

In `src/main/proto/greetings.proto`, we define:

* a `GreetingRequest` which holds a `name`, and
* a `GreetingReply` which holds a `message`

[source,protobuf]
.greetings.proto
----
include::src/main/proto/greetings.proto[]
----

=== Receiver verticle

The receiver verticle registers a consumer on the Event Bus.
When a request is received:

. the request is printed to the console along with its system _hashcode_
. a reply is generated
. the reply is printed to the console along with its system _hashcode_
. the reply is sent

[source,java]
.ReceiverVerticle.java
----
include::src/main/java/io/vertx/howtos/protobuf/eventbus/ReceiverVerticle.java[]
----

> NOTE: The system _hashcode_ is printed so that when we run the application in a single virtual machine, we can see whether objects are duplicated or not by the Event Bus.

Reviewer: ```suggestion
NOTE: The system _hash code_ is printed so that when we run the application in a single virtual machine, we can see whether objects are duplicated or not by the Event Bus.
```

[source,java]
.CustomLauncher.java
----
include::src/main/java/io/vertx/howtos/protobuf/eventbus/CustomLauncher.java[]
----

== Running the application

First you must build the application:

[source,shell]
----
./mvnw clean package
----

Then start the receiver:

[source,shell]
----
java -Djava.net.preferIPv4Stack=true -jar target/protobuf-eventbus-howto-1.0.0-SNAPSHOT-fat.jar run io.vertx.howtos.protobuf.eventbus.ReceiverVerticle -cluster
----

When it is ready, you will see: `INFO: Succeeded in deploying verticle`.

Now start the sender in another terminal:

[source,shell]
----
java -Djava.net.preferIPv4Stack=true -jar target/protobuf-eventbus-howto-1.0.0-SNAPSHOT-fat.jar run io.vertx.howtos.protobuf.eventbus.SenderVerticle -cluster
----

When it is ready, you will see: `INFO: Succeeded in deploying verticle`.

After some time, you will see in the sender console:

----
Sending request = Jane Doe (1445840961)
Received reply = Hello Jane Doe (654163465)
----

And in the receiver console:

----
Received request = Jane Doe (449456520)
Sending reply = Hello Jane Doe (522259462)
----

> In clustered mode, the system _hashcode_ that is printed is not important: objects living in distinct virtual machines are, obviously, different.

Reviewer: ```suggestion
In clustered mode, the system _hash code_ that is printed is not important: objects living in distinct virtual machines are, obviously, different.
```

What about local mode?
To run the sender and the receiver in the same virtual machine, we can use a third verticle whose only purpose is to deploy them.

[source,java]
.MainVerticle.java
----
include::src/main/java/io/vertx/howtos/protobuf/eventbus/MainVerticle.java[]
----

Open a terminal, build the project again and run the executable JAR.

[source,shell]
----
./mvnw clean package
java -jar target/protobuf-eventbus-howto-1.0.0-SNAPSHOT-fat.jar
----

When it is ready, you will see: `INFO: Succeeded in deploying verticle`.

After some time, you will see in the console:

----
Sending request = Jane Doe (346056258)
Received request = Jane Doe (346056258)
Sending reply = Hello Jane Doe (1483137857)
Received reply = Hello Jane Doe (1483137857)
----

> Pay attention to the system _hashcode_.

Reviewer: ```suggestion
Pay attention to the system _hash code_.
```
